### PR TITLE
[Fetch Migration] Added total document count to report

### PR DESCRIPTION
**Pull request by kartg**
### Description
This change allows the Index Configuration Tool report to print the total number of documents that will be migrated, based on the indices identified for creation. The change includes a new API call under `index_operations.py` and updated unit tests. 
* Category: Enhancement

This PR also includes some other minor changes:
1. The output YAML file is now optional, allowing users to print a report without producing a YAML file. However, one of `--report` or an output YAML file path are required - omitting both will result in a `ValueError`
2. A new `EndpointInfo` [dataclass](https://docs.python.org/3/library/dataclasses.html) has been introduced to encapsulate endpoint information (URL, auth and SSL verification flag) for source and target
3. The default output of the tool has been changed to dump the total document count followed by the list of created indices. `-r / --report` should be specified to produce human-readable output. All other print statements have been removed.

### Testing
```
$ python -m coverage report --omit "*/tests/*"
Name                  Stmts   Miss  Cover
-----------------------------------------
endpoint_info.py          6      0   100%
index_operations.py      35      0   100%
main.py                 121      1    99%
utils.py                 13      0   100%
-----------------------------------------
TOTAL                   175      1    99%
```

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: FetchMigration/index_configuration_tool/index_operations.py [hunk 1]
```diff
@@ -1,19 +1,20 @@
...
 __INDEX_KEY = "index"
 __ALL_INDICES_ENDPOINT = "*"
+__COUNT_ENDPOINT = "/_count"
 __INTERNAL_SETTINGS_KEYS = ["creation_date", "uuid", "provided_name", "version", "store"]
 
 
-def fetch_all_indices(endpoint: str, optional_auth: Optional[tuple] = None, verify: bool = True) -> dict:
-    actual_endpoint = endpoint + __ALL_INDICES_ENDPOINT
-    resp = requests.get(actual_endpoint, auth=optional_auth, verify=verify)
+def fetch_all_indices(endpoint: EndpointInfo) -> dict:
```
Comment (mikaylathompson) [inline FetchMigration/index_configuration_tool/index_operations.py:15]: ooh, I like this cleanup.

File: FetchMigration/index_configuration_tool/index_operations.py [hunk 2]
```diff
@@ -24,14 +25,21 @@ def fetch_all_indices(endpoint: str, optional_auth: Optional[tuple] = None, veri
...
         try:
-            resp = requests.put(actual_endpoint, auth=auth_tuple, json=data_dict)
+            resp = requests.put(actual_endpoint, auth=endpoint.auth, verify=endpoint.verify_ssl, json=data_dict)
             resp.raise_for_status()
         except requests.exceptions.RequestException as e:
-            print(f"Failed to create index [{index}] - {e!s}", file=sys.stderr)
+            raise RuntimeError(f"Failed to create index [{index}] - {e!s}")
+
+
+def doc_count(indices: set, endpoint: EndpointInfo) -> int:
```
Comment (mikaylathompson) [inline FetchMigration/index_configuration_tool/index_operations.py:41]: Should there be some try/catching in here?

Comment (mikaylathompson) [reply FetchMigration/index_configuration_tool/index_operations.py:41]: on the requests, but also casting to an int for a key that we're assumign is in a dict is always risky

File: FetchMigration/index_configuration_tool/main.py [hunk 3]
```diff
@@ -138,44 +134,64 @@ def get_index_differences(source: dict, target: dict) -> tuple[set, set, set]:
...
     print("Identical indices in the target cluster (no changes will be made): " +
           utils.string_from_set(index_differences[1]))
     print("Indices in target cluster with conflicting settings/mappings: " +
           utils.string_from_set(index_differences[2]))
     print("Indices to create: " + utils.string_from_set(index_differences[0]))
+    print("Total documents to be moved: " + str(count))
+
+
+def dump_count_and_indices(count: int, indices: set):  # pragma no cover
+    print(count)
```
Comment (mikaylathompson) [inline FetchMigration/index_configuration_tool/main.py:147]: Can we put a bit of contextual information around this? Like a label before we print it?

**Review by mikaylathompson — APPROVED** [review]

File: FetchMigration/index_configuration_tool/index_operations.py (hunk 2 above)
Comment (kartg) [reply FetchMigration/index_configuration_tool/index_operations.py:41]: The `count` member of the [_count API](https://www.elastic.co/guide/en/elasticsearch/reference/7.10/search-count.html) will always be an int so I think that part of it is safe.

I'll add a backlog item to improve try-catching on requests throughout the FetchMigration tool (they're all sadly lacking)

**Pull request merged**